### PR TITLE
feat(preprocess): Function hoisting

### PR DESCRIPTION
**Pull request by EdenComp**
Also fixes GLA-136

### CODE REVIEW

File: src/Dreamberd/Compilation/Preprocessing.hs [hunk 1]
```diff
@@ -4,12 +4,19 @@ module Dreamberd.Compilation.Preprocessing (
...
+import Data.List (partition)
 import Dreamberd.Parser (parseDreamberd)
 import Dreamberd.Types (AstNode (..), File (..))
 import System.Directory (getCurrentDirectory, getHomeDirectory)
 
 executePreprocessing :: File [AstNode] -> IO (Either String [AstNode])
-executePreprocessing (File filename nodes) = do
+executePreprocessing file =
+    preprocessImports file >>= \case
+        Right nodes' -> return $ Right $ preprocessFunctions nodes'
```
Comment (Croos3r) [inline src/Dreamberd/Compilation/Preprocessing.hs:15]: omg double monads

Comment (EdenComp) [reply src/Dreamberd/Compilation/Preprocessing.hs:15]: IO (Either String (Maybe IO Int))

File: src/Dreamberd/Compilation/Preprocessing.hs [hunk 2]
```diff
@@ -69,3 +76,16 @@ importFile :: String -> ([String], String) -> IO (Either String ([String], [AstN
...
     Left err -> return $ Left ("Error while importing " ++ lib ++ ": " ++ err)
+
+preprocessFunctions :: [AstNode] -> [AstNode]
+preprocessFunctions a = concatMap hoistFunction functions ++ rest
+  where
+    (functions, rest) = partition isFunction a
+
+isFunction :: AstNode -> Bool
+isFunction Function{} = True
+isFunction _ = False
```
Comment (Croos3r) [inline src/Dreamberd/Compilation/Preprocessing.hs:87]: You could declare this inside the preprocessFunctions as it is not used anywhere else

File: src/Dreamberd/Compilation/Preprocessing.hs [hunk 3]
```diff
@@ -69,3 +76,16 @@ importFile :: String -> ([String], String) -> IO (Either String ([String], [AstN
...
+
+hoistFunction :: AstNode -> [AstNode]
+hoistFunction (Function name args ast) = [Function name args (preprocessFunctions ast)]
```
Comment (Croos3r) [inline src/Dreamberd/Compilation/Preprocessing.hs:90]: You are missing If and Loop

**Review by Croos3r — CHANGES_REQUESTED** [review]

File: src/Dreamberd/Compilation/Preprocessing.hs (hunk 2 above)
Comment (EdenComp) [reply src/Dreamberd/Compilation/Preprocessing.hs:87]: I need pattern matching,using a lambda here would require a case

File: src/Dreamberd/Compilation/Preprocessing.hs (hunk 3 above)
Comment (EdenComp) [reply src/Dreamberd/Compilation/Preprocessing.hs]: Fixed

File: src/Dreamberd/Compilation/Preprocessing.hs (hunk 2 above)
Comment (Croos3r) [reply src/Dreamberd/Compilation/Preprocessing.hs:87]: You can define functions in the where of another function like in the parseChar of Parser.hs

**Review by Croos3r — APPROVED** [review]
LGTM

**Pull request merged**